homeassistant/components/screenlogic/config_flow.py

```diff
@@ -0,0 +1,247 @@
+"""Config flow for ScreenLogic."""
+import logging
+
+from screenlogicpy import ScreenLogicError, discover
+from screenlogicpy.const import SL_GATEWAY_IP, SL_GATEWAY_NAME, SL_GATEWAY_PORT
+from screenlogicpy.requests import login
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.dhcp import HOSTNAME, IP_ADDRESS
+from homeassistant.const import CONF_IP_ADDRESS, CONF_PORT, CONF_SCAN_INTERVAL
+from homeassistant.core import callback
+from homeassistant.helpers import config_entry_flow
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.device_registry import format_mac
+
+from .const import DEFAULT_SCAN_INTERVAL, DOMAIN, MIN_SCAN_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+GATEWAY_SELECT_KEY = "selected_gateway"
+GATEWAY_MANUAL_ENTRY = "manual"
+
+GATEWAY_ENTRY_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_IP_ADDRESS): str,
+        vol.Required(CONF_PORT, default=80): int,
+    }
+)
+
+PENTAIR_OUI = "00-C0-33"
+
+
+async def async_discover_gateways_by_unique_id(hass):
+    """Discover gateways and return a dict of them by unique id."""
+    discovered_gateways = {}
+    try:
+        hosts = await hass.async_add_executor_job(discover)
+        _LOGGER.debug("Discovered hosts: %s", hosts)
+    except ScreenLogicError as ex:
+        _LOGGER.debug(ex)
+        return discovered_gateways
+
+    for host in hosts:
+        mac = _extract_mac_from_name(host[SL_GATEWAY_NAME])
+        discovered_gateways[mac] = host
+
+    _LOGGER.debug("Discovered gateways: %s", discovered_gateways)
+    return discovered_gateways
+
+
+def _extract_mac_from_name(name):
+    return format_mac(f"{PENTAIR_OUI}-{name.split(':')[1].strip()}")
+
+
+def short_mac(mac):
+    """Short version of the mac as seen in the app."""
+    return "-".join(mac.split(":")[3:]).upper()
+
+
+def name_for_mac(mac):
+    """Derive the gateway name from the mac."""
+    return f"Pentair: {short_mac(mac)}"
+
+
+async def _async_has_devices(hass):
+    """Return if there are devices that can be discovered."""
+    return bool(await async_discover_gateways_by_unique_id(hass))
+
+
+_LOGGER.info("Registering discovery flow")
+config_entry_flow.register_discovery_flow(
+    DOMAIN,
+    "Pentair ScreenLogic",
+    _async_has_devices,
+    config_entries.CONN_CLASS_LOCAL_POLL,
+)
```

Comment:
Not sure this is needed since we have dhcp discovery below
```suggestion
```

Reply:
Will take a look.

Reply:
Need to double check on this though in the morning after some sleep